File: DownloadModule/prepare_code_change/collect-recent-graph.js
const Mongoose = require('mongoose');
const Moment = require('moment');
const MathJs = require('mathjs');
const jsnx = require('jsnetworkx');
const cliProgress = require('cli-progress');
const Database = require('../config/databaseConfig');
const Change = require('../models/change');
const Metrics = require('../models/metrics');
const MetricsUtils = require('../compute_metrics/metrics-utils');
const Utils = require('../config/utils');
const ApiEndPoints = require('../config/apiEndpoints');
const PathLibrary = require('path');

const progressBar = new cliProgress.SingleBar({
    barCompleteChar: '#',
    barIncompleteChar: '-',
}, cliProgress.Presets.shades_classic);

let projectJson = Utils.getProjectParameters("libreoffice");
let projectDBUrl = projectJson["projectDBUrl"];
let projectApiUrl = projectJson["projectApiUrl"];
let projectName = projectJson["projectName"];
let DATA_PATH = "data/";

let STARTING_POINT = 0;
let NUM_DAYS_FOR_RECENT = 90;
let NUM_OF_CHANGES_LIMIT = 10000;
let NUMBER_DATABASE_REQUEST = Utils.getCPUCount() ? Utils.getCPUCount() : 4;
//let overAllGraphJson = {};
//let overAllFullConnectedGraphJson = {};
//let overAllChangesAccountInfo = {};
let changes_graph_list = {}

let i = 0;

//do reduce graph
//collect change in graph folder by number an compare to the previous

if (typeof require !== 'undefined' && require.main === module) {
    start(projectJson)
        .catch(err => {
            console.log(err)
        });
}

function start(json) {
    if (json["projectDBUrl"])
        projectDBUrl = json["projectDBUrl"];
    if (json["projectApiUrl"])
        projectApiUrl = json["projectApiUrl"];
    if (json["output_directory"])
        DATA_PATH = json["output_directory"];
    if (json["projectName"])
        projectName = json["projectName"];
    if (json["numberOfDays"])
        NUM_DAYS_FOR_RECENT = json["numberOfDays"];

    console.log("Collecting account graph !!!!");

    return Database.dbConnection(projectDBUrl)
        .then(() => { // Counts the number of change
            return Change.estimatedDocumentCount({});
        })
        .then((count) => {
            //NUM_OF_CHANGES_LIMIT = MathJs.ceil(count / NUMBER_DATABASE_REQUEST);
            console.log("Processing data by slice of " + NUM_OF_CHANGES_LIMIT);
            progressBar.start(count, 0);
            return getChanges(STARTING_POINT);
        })
        .then(() => {
            /*let name1 = projectName + "-graph";
            let name2 = projectName + "-full-connected-graph";
            let name3 = projectName + "-changes-account-info";
            let path = PathLibrary.join(DATA_PATH, projectName);
            let t1 = Utils.saveJSONInFile(path, name1, overAllGraphJson);
            let t2 = Utils.saveJSONInFile(path, name2, overAllFullConnectedGraphJson);
            let t3 = Utils.saveJSONInFile(path, name3, overAllChangesAccountInfo);
            return Promise.all([t1, t2, t3]);*/

            let name = projectName + "-changes-graph-list-" + NUM_DAYS_FOR_RECENT + "-days";
            let path = PathLibrary.join(DATA_PATH, projectName);
            return Utils.saveJSONInFile(path, name, changes_graph_list);
        })
        .then(() => {
            progressBar.stop();
            console.log("Finished!!!!");
            return Database.closeConnection();
        })
        .catch(err => {
            console.log(err)
        });
}

function getChanges(skip) {
    return Change
        .aggregate([
            {$sort: {_number: 1}},
            {
                $project: {
                    id: 1,
                    created: 1,
                    updated: 1,
                    _number: 1,
                    owner_id: "$owner._account_id",
                    reviewers_id: "$reviewers.REVIEWER._account_id"
                }
            },
            {$skip: skip},
            {$limit: NUM_OF_CHANGES_LIMIT}
        ])
        .allowDiskUse(true)
        .exec()
        .then(docs => {
            return docs.length ? collectDocs(docs) : Promise.resolve(false);
        })
        .then(result => {
            return result ? getChanges(skip + NUM_OF_CHANGES_LIMIT) : Promise.resolve(false);
        })
        .catch(err => {
            console.log(err)
        });
}

async function collectDocs(docs) {
    if (!docs)
        return Promise.resolve(true);
    let previousJson = {};
    let graph_list = {};
    for (let key in docs) {
        let previousGraph = await collectGraph(docs[key], previousJson, graph_list)
            .then((result) => {
                updateProgress();
                //console.log(result)
                return Promise.resolve(result);
            });
        previousJson = docs[key];
        graph_list = previousGraph;
        //console.log("echo")
    }
    return Promise.resolve(true);
}

function updateProgress() {
    progressBar.increment(1);
}

//let graph_list = {}
//let previous_graph_list = {}

function buildChangeAccountInfoJson(json) {
    let id = json.id
    let owner_id = json.owner_id;
    let number = json._number;
    let reviewers_id = json.reviewers_id;
    return {id: id, number: number, owner_id: owner_id, reviewers_id: reviewers_id}
}


async function collectGraph(json, previousJson, graph_list) {
    return getIntermediaryUpdatedChanges(json, previousJson)
        .then((hasChanged) => {
            if (hasChanged) {
                return getPriorChangesGraph(json, previousJson, graph_list);
            } else {
                let id = json.id
                //let changeAccountInfoJson = buildChangeAccountInfoJson(json);
                //let path = PathLibrary.join(DATA_PATH, projectName, "changes-account");
                //let t1 = Utils.saveJSONInFile(path, id, changeAccountInfoJson);
                /*if (graph_list["changes"])
                    graph_list["changes"][id] = changeAccountInfoJson;*/
                changes_graph_list[id] = i;
                //let t2 = Promise.resolve(graph_list);
                return Promise.resolve(graph_list);
                /*return Promise.all([t1, t2])
                    .then((results) => {
                        return Promise.resolve(results[1])
                    });*/
            }
        })
}
function getPriorChangesGraph(json, previousJson, graph_list){
    return getPriorChanges(json)
        .then((results) => {
            let id = json.id
            let owner_id = json.owner_id;
            //let changeAccountInfoJson = buildChangeAccountInfoJson(json);
            //let path = PathLibrary.join(DATA_PATH, projectName, "changes-account");
            let t1 = buildGraph(results, id, owner_id);
            //let t2 = buildFullConnectedGraph(results, id, owner_id);
            let t3 = getIntermediaryUpdatedChanges(json, previousJson)
            //let t4 = Promise.resolve(changeAccountInfoJson);
            //let t5 = Utils.saveJSONInFile(path, id, changeAccountInfoJson);
            //return Promise.all([t1, t2, t3, t4, t5]);
            //return Promise.all([t1, t2, t3, t4]);
            //return Promise.all([t1, t2, t3]);
            return Promise.all([t1, t3]);
        })
        .then((results) => {
            let id = json.id;
            let graph = results[0];
            //let fullConnectedGraph = results[1];
            let hasChanged = results[1];
            //let changeAccountInfoJson = results[3];
            let suffix = "-" + NUM_DAYS_FOR_RECENT + "-days";
            //let path1 = PathLibrary.join(DATA_PATH, projectName, "changes-graph" + suffix);
            //let path2 = PathLibrary.join(DATA_PATH, projectName, "changes-full-connected-graph" + suffix);

            //let t1 = Utils.saveJSONInFile(path1, id, graph);
            //let t2 = Utils.saveJSONInFile(path2, id, fullConnectedGraph);
            //overAllGraphJson[id] = graph;
            //overAllFullConnectedGraphJson[id] = fullConnectedGraph;

            //compare to the previous graph
            let path3 = PathLibrary.join(DATA_PATH, projectName, "graph-list" + suffix);
            let t3 = Promise.resolve(true);
            if (hasChanged) {

                if (i > 0) {
                    t3 = Utils.saveJSONInFile(path3, String(i - 1), graph_list);
                }
                i++;

                graph_list = {};
                graph_list["graph"] = graph;
                //graph_list["full_connected_graph"] = fullConnectedGraph;
                changes_graph_list[id] = i;
                //graph_list["changes"] = {};
                //graph_list["changes"][id] = changeAccountInfoJson;

            } else {
                //if (graph_list["changes"])
                //    graph_list["changes"][id] = changeAccountInfoJson;
                changes_graph_list[id] = i;
            }
            let t4 = Promise.resolve(graph_list)
            //return Promise.all([t1, t2, t3, t4]);
            return Promise.all([t3, t4]);
        })
        .then((results) => {
            return Promise.resolve(results[1])
        })
        .catch(err => {
            console.log(err)
        });
}

async function getIntermediaryUpdatedChanges(json, previousJson) {
    if (!previousJson)
        return true;

    if (!previousJson.created)
        return true;

    let created = json.created;
    let previous_created = previousJson.created;
    let created_minus_recent_days = Moment(json.created).subtract(NUM_DAYS_FOR_RECENT, 'days').format('YYYY-MM-DD HH:mm:ss.SSSSSSSSS');
    let previous_created_minus_recent_days = Moment(previousJson.created).subtract(NUM_DAYS_FOR_RECENT, 'days').format('YYYY-MM-DD HH:mm:ss.SSSSSSSSS');
    let pipeline = [
        {
            $match: {
                status: {$in: ['MERGED', 'ABANDONED']},
                $or: [
                    {updated: {$lte: created, $gte: previous_created}},
                    {updated: {$lte: created_minus_recent_days, $gte: previous_created_minus_recent_days}},
                ]
            }
        },
        {$count: "count"},
    ];
    return Change
        .aggregate(pipeline)
        .allowDiskUse(true)
        .exec()
        .then(docs => {
            //hasChanged?
            if (!docs)
                return false;
            if (docs.length > 0) {
                return docs[0].count > 0;
            } else {
                return false;
            }
            //return docs.length > 0 ? (docs[0].count > 0) : false;
        })
        .catch(err => {
            console.log(err)
        });
}

function getPriorChanges(json) {
    let priorResults = [[], []];
    return getPriorChangesFromDB(0, json, priorResults);
}

function getPriorChangesFromDB(skip, json, priorResults) {
    let number = json._number;
    let endDate = json.created;
    let startDate = Moment(json.created).subtract(NUM_DAYS_FOR_RECENT, 'days').format('YYYY-MM-DD HH:mm:ss.SSSSSSSSS');
    let pipeline = [
        {
            $match: {
                status: {$in: ['MERGED', 'ABANDONED']},
                updated: {$lte: endDate},
                created: {$lte: endDate, $gte: startDate},
                _number: {$lt: number},
            }
        },
        {$project: {id: 1, owner_id: "$owner._account_id", reviewers_id: "$reviewers.REVIEWER._account_id"}},
        {$skip: skip},
        {$limit: 2000}
    ];
    //console.log("startDate : " + startDate);
    //console.log("endDate : " + endDate);
    return dbRequest(skip, json, pipeline, priorResults);
}

function dbRequest(skip, json, pipeline, previousResults) {
    return Change
        .aggregate(pipeline)
        .allowDiskUse(true)
        .exec()
        .then(docs => {
            if (!docs)
                return Promise.resolve([]);
            return docs.length > 0 ? Promise.resolve(docs) : Promise.resolve(true);
        })
        .then(result => {
            if (typeof result === "boolean") {
                return Promise.resolve(previousResults)
            } else {
                previousResults.push(...result)
                return getPriorChangesFromDB(skip + 1000, json, previousResults)
            }
        })
        .catch(err => {
            console.log(err)
        });
}

async function buildGraph(docs, id, owner) {
    let nodes = new Set();
    let edges = [];
    for (let i = 0; i < docs.length; i++) {
        let doc = docs[i];
        let owner_id = doc["owner_id"];
        if (!owner_id) continue;
        nodes.add(owner_id);
        let reviewersId = doc["reviewers_id"];
        if (reviewersId)
            for (let j = 0; j < reviewersId.length; j++) {
                let rev_id = reviewersId[j];
                //console.log(reviewersId[j]);
                if (!rev_id) continue;
                if (MetricsUtils.isABot(rev_id, projectName)) continue;
                nodes.add(rev_id);
                let edge = [owner_id, rev_id];
                edges.push(edge);
            }
    }
    let graph = {};
    //graph["id"] = id;
    //graph["owner_id"] = owner;
    graph["nodes"] = [...nodes];
    graph["edges"] = [...edges];
    return graph;
}

/*Array.prototype.except = function (val) {
    return this.filter(function (x) {
        return x !== val;
    });
};*/

let bot_filter = function (account_id) {
    return !MetricsUtils.isABot(account_id, projectName);
};

function excludeBot(array) {
    return array.filter(function (account_id) {
        return !MetricsUtils.isABot(account_id, projectName);
    });
}


function buildFullConnectedGraph(docs, id, owner) {
    let nodes = new Set();
    let edges = [];
    for (let i = 0; i < docs.length; i++) {
        let doc = docs[i];
        let owner_id = doc["owner_id"];
        if (!owner_id) continue;
        nodes.add(owner_id);
        let reviewersId = doc["reviewers_id"];
        if (reviewersId)
            for (let j = 0; j < reviewersId.length; j++) {
                let revId1 = reviewersId[j];
                if (MetricsUtils.isABot(revId1, projectName)) continue;
                nodes.add(revId1);
                let edge = [owner_id, revId1];
                edges.push(edge);

                for (let k = 0; k < reviewersId.length; k++) {
                    let revId2 = reviewersId[k];
                    if (MetricsUtils.isABot(revId2, projectName)) continue;
                    if (revId1 === revId2) continue;
                    let edge = [revId1, revId2];
                    edges.push(edge);
                }

            }
    }
    let graph = {};
    //graph["id"] = id;
    //graph["owner_id"] = owner;
    graph["nodes"] = [...nodes];
    graph["edges"] = [...edges];
    return graph;
}

/*let sortedArrayFunction = function(a, b) {
    if (a[0] == b[0]) {
        return a[1] - b[1];
    }
    return a[0] - b[0];
};*/

let sortedArrayFunction = ([a, b], [c, d]) => a - c || d - b;

function compareGraph(first, second) {
    if (!first.nodes && !second.nodes && !first.edges && !second.edges)
        return false;

    let firstNodes = first.nodes.sort();
    let secondNodes = second.nodes.sort();
    let firstEdges = first.edges.sort(sortedArrayFunction);
    let secondEdges = second.edges.sort(sortedArrayFunction);
    let areNodesEqual = firstNodes.length === secondNodes.length && firstNodes.every((value, index) => value === secondNodes[index])
    let areEdgesEqual = firstEdges.length === secondEdges.length && firstEdges.every(
        (value, index) => JSON.stringify(value) === JSON.stringify(secondEdges[index]))

    return areNodesEqual && areEdgesEqual;
}

module.exports = {
    start: start
};

File: MetricsAnalysisModule/collect-recent-graph.py
import dbutils
import utils
import json
import os
import moment
from datetime import datetime
import networkx as nx 
import sys
import code_metrics
from networkx.readwrite import json_graph
from pathlib import Path as pathlib
from utils import SlowBar as SlowBar

DATA_DIR_NAME = "/Volumes/SEAGATE-II/Data/libreoffice/"
GRAPHS_GPICKLE_PATH = '/Volumes/SEAGATE-II/server-data-sync/libreoffice/graph-gpickle-180-days'
FULL_GRAPHS_GPICKLE_PATH = '/Volumes/SEAGATE-II/server-data-sync/libreoffice/graph-full-gpickle-180-days'
CHANGES_GRAPH_LIST_PATH = '/Volumes/SEAGATE-II/server-data-sync/libreoffice/changes-graph-list-180-days.json'

'''
JSON_GRAPHS_GPICKLE_PATH = '/Volumes/SEAGATE-II/server-data-sync/libreoffice/graph-json'
JSON_FULL_GRAPHS_GPICKLE_PATH = '/Volumes/SEAGATE-II/server-data-sync/libreoffice/graph-full-json'
'''

PROJET_NAME = "libreoffice"
BOT_ACCOUNT = utils.get_bot_accounts(PROJET_NAME)
NUM_DAYS_FOR_RECENT = 180

Database = None
bar = None
changes_graph_list = {}
changes = {}
i = 0


def start(json):
    
    global PROJET_NAME
    PROJET_NAME = json["project_name"]

    global DATA_DIR_NAME
    DATA_DIR_NAME = json["output_data_path"]

    global GRAPHS_GPICKLE_PATH
    GRAPHS_GPICKLE_PATH = os.path.join(DATA_DIR_NAME,PROJET_NAME, "graph-gpickle")
    
    global FULL_GRAPHS_GPICKLE_PATH
    FULL_GRAPHS_GPICKLE_PATH = os.path.join(DATA_DIR_NAME,PROJET_NAME, "graph-full-gpickle")
    
    global CHANGES_GRAPH_LIST_PATH
    CHANGES_GRAPH_LIST_PATH = os.path.join(DATA_DIR_NAME,PROJET_NAME, PROJET_NAME + "-changes-graph-list-" + str(NUM_DAYS_FOR_RECENT) + "-days.json")

    pathlib(GRAPHS_GPICKLE_PATH).mkdir(parents=True, exist_ok=True)
    pathlib(FULL_GRAPHS_GPICKLE_PATH).mkdir(parents=True, exist_ok=True)

    global Database
    Database = dbutils.getDatabaseFromJson(json)
    
    global count
    count = Database.get_changes_count()
    
    global bar
    bar = SlowBar('Collecting graph', max=count)

    collect_changes(0)
    return 0


def collect_changes(skip):
    changes = Database.get_changes_list_for_graph(skip)
    previous_json = {}
    graph = nx.Graph()
    full_connected_graph = nx.Graph()
    if len(changes) > 0:
        for doc in changes:
            result = collect_graph(doc, previous_json, graph, full_connected_graph)
            previous_json = doc
            graph = result[0]
            full_connected_graph = result[1]
            bar.next()
        del changes
        return collect_changes(skip + dbutils.NUM_OF_CHANGES_LIMIT)
    print("\nFinish")
    save_json(changes_graph_list, CHANGES_GRAPH_LIST_PATH)
    bar.finish()
    return 0


def save_json(json_graph, path):
    dir_path = pathlib(path)
    dir_path.parent.mkdir(parents=True, exist_ok=True)
    with open(path, "wb") as f:
        f.write(json.dumps(json_graph, indent=4).encode("utf-8"))
        f.close()
    return 0


def collect_graph(json, previous_json, graph, full_connected_graph):
    added_changes = getAddedIntermediaryUpdatedChanges(json, previous_json)
    removed_changes = getRemovedIntermediaryUpdatedChanges(json, previous_json)    
    
    if (len(added_changes)>0 or len(removed_changes)>0) :
        return updated_changes_in_graph(json, added_changes, removed_changes, graph, full_connected_graph)
    else:
        global i
        id = json["id"]
        changes_graph_list[id] = i
        if(i == 0):
            graph = nx.Graph()
            full_connected_graph = nx.Graph()
            save_graph(graph, full_connected_graph, str(i))
            i+=1
        return [graph, full_connected_graph]


def getAddedIntermediaryUpdatedChanges(json, previous_json):
    if(not previous_json):
        return []
    if (not "created" in previous_json):
        return []
    created = json["created"]
    previous_created = previous_json["created"]
    pipeline = [
        {
            "$match": {
                "status": {"$in": ['MERGED', 'ABANDONED']},
                "updated": {"$lte": created, "$gte": previous_created}
            }
        },
        {"$project": {"id": 1, "owner_id": "$owner._account_id", "reviewers_id": "$reviewers.REVIEWER._account_id"}},
    ]
    changes_collection = Database.get_changes_collection()
    return list(changes_collection.aggregate(pipeline, allowDiskUse=True))


def getRemovedIntermediaryUpdatedChanges(json, previous_json):
    if(not previous_json):
        return []
    if (not "created" in previous_json):
        return []
    created_minus_recent_days = moment.date(json["created"]).subtract(days=NUM_DAYS_FOR_RECENT).format('YYYY-MM-DD HH:mm:ss.000000000');
    previous_created_minus_recent_days = moment.date(previous_json["created"]).subtract(days=NUM_DAYS_FOR_RECENT).format('YYYY-MM-DD HH:mm:ss.000000000');
    pipeline = [
        {
            "$match": {
                "status": {"$in": ['MERGED', 'ABANDONED']},
                "updated": {"$lte": created_minus_recent_days, "$gte": previous_created_minus_recent_days}
            }
        },
        {"$project": {"id": 1, "owner_id": "$owner._account_id", "reviewers_id": "$reviewers.REVIEWER._account_id"}},
    ]
    changes_collection = Database.get_changes_collection()
    return list(changes_collection.aggregate(pipeline, allowDiskUse=True))


def update_changes(added_changes, removed_changes):
    for change in added_changes:
        id = change["id"]
        changes[id] = change
    for change in removed_changes:
        id = change["id"]
        changes.pop(id)

        
def updated_changes_in_graph(json, added_changes, removed_changes, graph, full_connected_graph):
    update_changes(added_changes, removed_changes)
    id = json["id"]
    owner_id = json["owner_id"]
    global i
    if (i > 0):
        save_graph(graph, full_connected_graph, str(i-1))
    i+=1
    changes_graph_list[id] = i
    graph = build_graph(changes, owner_id);
    full_connected_graph = build_full_connected_graph(changes, owner_id);
    return[graph, full_connected_graph]


def save_graph(G, FG, filename):
    graph_path = GRAPHS_GPICKLE_PATH + "-" + str(NUM_DAYS_FOR_RECENT) + "days"
    full_graph_path = FULL_GRAPHS_GPICKLE_PATH + "-" + str(NUM_DAYS_FOR_RECENT) + "days"
    pathForSimple = os.path.join(graph_path, filename + ".gpickle")
    pathForFullSimple = os.path.join(full_graph_path, filename + ".gpickle")
    nx.write_gpickle(G, pathForSimple)
    nx.write_gpickle(FG, pathForFullSimple)

    '''
    pathForSimpleJson = os.path.join(JSON_GRAPHS_GPICKLE_PATH + "-" + str(NUM_DAYS_FOR_RECENT) + "days", filename)
    pathForFullSimpleJson = os.path.join(JSON_FULL_GRAPHS_GPICKLE_PATH + "-" + str(NUM_DAYS_FOR_RECENT) + "days", filename)
    save_graph_json(G, pathForSimpleJson)
    save_graph_json(FG, pathForFullSimpleJson)
    '''

    return 0


def save_graph_json(G, fname):
    graph = dict()
    graph["graph"] = {}
    graph["graph"]["nodes"] = [n for n in G.nodes()]
    graph["graph"]["edges"] = []
    for u,v in G.edges():
        for i in range(G[u][v]['weight']):
            graph["graph"]["edges"].append([u,v])
    json.dump(graph, open(fname + ".json", 'w'), indent=2)


def build_graph(changes, owner_id):
    graph = nx.Graph()
    for key in changes:
        doc = changes[key]
        owner_id = doc["owner_id"]
        if ("reviewers_id" in doc):
            reviewers_id = exclude_bot(doc["reviewers_id"])
        else:
            reviewers_id = []

        graph.add_node(owner_id)
        graph.add_nodes_from(reviewers_id)
        for rev_id in reviewers_id:
            if graph.has_edge(owner_id, rev_id):
                graph[owner_id][rev_id]['weight'] += 1
            else:
                graph.add_edge(owner_id, rev_id, weight=1)
    graph.remove_edges_from(nx.selfloop_edges(graph))
    return graph


def build_full_connected_graph(changes, owner_id):
    graph = nx.Graph()
    for key in changes:
        doc = changes[key]
        owner_id = doc["owner_id"]
        if ("reviewers_id" in doc):
            reviewers_id = exclude_bot(doc["reviewers_id"])
        else:
            reviewers_id = []
        graph.add_node(owner_id)
        graph.add_nodes_from(reviewers_id)
        for rev_id in reviewers_id:
            if graph.has_edge(owner_id, rev_id):
                graph[owner_id][rev_id]['weight'] += 1
            else:
                graph.add_edge(owner_id, rev_id, weight=1)

            for rev_id2 in reviewers_id:
                if(rev_id == rev_id2):
                    continue
                if graph.has_edge(rev_id, rev_id2):
                    graph[rev_id][rev_id2]['weight'] += 1
                else:
                    graph.add_edge(rev_id, rev_id2, weight=1)
    graph.remove_edges_from(nx.selfloop_edges(graph))
    return graph


def exclude_bot(reviewers_id):
    return [a for a in reviewers_id if a not in BOT_ACCOUNT]


if __name__ == '__main__':
    '''
    pathlib(GRAPHS_GPICKLE_PATH + "-" + str(NUM_DAYS_FOR_RECENT) + "days").mkdir(parents=True, exist_ok=True)
    pathlib(FULL_GRAPHS_GPICKLE_PATH + "-" + str(NUM_DAYS_FOR_RECENT) + "days").mkdir(parents=True, exist_ok=True)
    pathlib(JSON_GRAPHS_GPICKLE_PATH + "-" + str(NUM_DAYS_FOR_RECENT) + "days").mkdir(parents=True, exist_ok=True)
    pathlib(JSON_FULL_GRAPHS_GPICKLE_PATH + "-" + str(NUM_DAYS_FOR_RECENT) + "days").mkdir(parents=True, exist_ok=True)
    collect_changes(0)
    '''
    argument = sys.argv[1:]

    if(argument):
        projectName = argument[0]
        projectJson = code_metrics.get_project_json(projectName)
        if(projectJson is not None):
            start(projectJson)
        else:
            print ("The project you request can't not be found in the Config.json file")
    else:
        print ("Please provide an argument")
